### PR TITLE
Add new pre-commit checks

### PR DESCRIPTION
**Pull request by mraspaud**
This PR add checks for cyclomatic complexity, security and typing.

I feel this is highly overdue...

Cyclomatic complexity: I think the maintainers of satpy have accepted that clean coding has to be enforced in order to have maintainable code. Cyclomatic complexity measure how complex a piece of code is and thus helps flag for organically growing functions for example. The threshold is now 10, which I think is a reasonable with our current standards in the new code I have seen being added to satpy in the last year.

Security: Several institution/individuals are starting to install and run satpy in the cloud. As maintainers, we have to make sure that the satpy library is safe to have running in any kinds of environments and is resistant against common kinds of attacks. For me this includes ingestion of tempered satellite data, so it feels relevant to have a tool such as `bandit` running regularly.

Typing: We have started using type annotations in satpy, hence it feels natural to run mypy on it.

 - [ ] Fully documented

### CODE REVIEW

**Comment by djhoese** [comment]
isort needs some extra configuration in setup.cfg:

https://github.com/pytroll/pyresample/blob/e6a452ece4e5c4f7609136a20795cf3f0cef388a/setup.cfg#L26-L32

Also, this is going to be a difficult PR to get passing as pre-commit.ci must run on the entire code base.

**Comment by mraspaud** [comment]
> isort needs some extra configuration in setup.cfg:
> 
> https://github.com/pytroll/pyresample/blob/e6a452ece4e5c4f7609136a20795cf3f0cef388a/setup.cfg#L26-L32

thanks

> Also, this is going to be a difficult PR to get passing as pre-commit.ci must run on the entire code base.

Fun :)

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -1,8 +1,26 @@
...
 fail_fast: false
 repos:
--   repo: https://github.com/pre-commit/pre-commit-hooks
+  - repo: https://github.com/pre-commit/pre-commit-hooks
     rev: v2.2.3
     hooks:
     - id: flake8
-      additional_dependencies: [flake8-docstrings, flake8-debugger, flake8-bugbear]
+      additional_dependencies: [flake8-docstrings, flake8-debugger, flake8-bugbear, mccabe]
+      args: [--max-complexity, "10"]
```
Comment (pnuu) [inline .pre-commit-config.yaml:9]: Should these be also updated to `.pre-commit-config.yaml` so that they are run before a commit is allowed?

Comment (djhoese) [reply .pre-commit-config.yaml:9]: This is .pre-commit-config.yaml

Comment (pnuu) [reply .pre-commit-config.yaml:9]: :man_facepalming:

**Comment by djhoese** [comment]
So I just pushed a couple commits. First, apparently the `force_on_top` option is supposed to be a set of import names that should be forced to the top and not a boolean. This wasn't obvious until using pyproject.toml for the config I guess.

Next I updated the isort line length to 120 to match flake8. It was trying to shorten too many imports in the package.

Lastly, I forced the modifier import order so that the base modifier is always imported first to avoid circular imports.

File: satpy/demo/_google_cloud_platform.py [hunk 2]
```diff
@@ -24,7 +24,7 @@
     from urllib.request import urlopen
 except ImportError:
     # python 2
-    from urllib2 import URLError, urlopen
+    from urllib2 import URLError, urlopen  # type: ignore
```
Comment (mraspaud) [inline satpy/demo/_google_cloud_platform.py:27]: We are not supporting python2 anymore, right?

Comment (djhoese) [reply satpy/demo/_google_cloud_platform.py:27]: Ah yes, good catch. I missed that. I'll update that.

**Comment by sfinkens** [comment]
:+1: Complexity threshold of 10 is a good choice I think. Thanks for starting the refactoring already! Happy to help next week if there's something left to refactor.

File: satpy/demo/__init__.py [hunk 3]
```diff
@@ -50,8 +50,9 @@
 
 """
 
-from .abi_l1b import get_hurricane_florence_abi, get_us_midlatitude_cyclone_abi  # noqa
-from .ahi_hsd import download_typhoon_surigae_ahi  # noqa
-from .viirs_sdr import get_viirs_sdr_20170128_1229  # noqa
-from .fci import download_fci_test_data  # noqa
-from .seviri_hrit import download_seviri_hrit_20180228_1500  # noqa
+from .abi_l1b import get_hurricane_florence_abi  # noqa: F401
+from .abi_l1b import get_us_midlatitude_cyclone_abi  # noqa: F401
```
Comment (gerritholl) [inline satpy/demo/__init__.py:54]: Why did this change from one to two lines for importing two names from one module?

File: satpy/demo/fci.py [hunk 4]
```diff
@@ -17,12 +17,14 @@
...
-import tarfile
 import pathlib
+import tarfile
 import tempfile
-from . import utils
+
 from satpy import config
 
+from . import utils
+
```
Comment (gerritholl) [inline satpy/demo/fci.py:27]: Are the rules for import orders written down somewhere?  The changes on the import order seems to go beyond what is formulated in PEP 8.

File: satpy/modifiers/__init__.py [hunk 5]
```diff
@@ -19,7 +19,11 @@
...
 # file deepcode ignore W0611: Ignore unused imports in init module
 
-from .base import ModifierBase  # noqa: F401
-from .spectral import NIREmissivePartFromReflectance, NIRReflectance  # noqa: F401
-from .geometry import SunZenithCorrector, EffectiveSolarPathLengthCorrector  # noqa: F401
-from .atmosphere import PSPRayleighReflectance, PSPAtmosphericalCorrection, CO2Corrector  # noqa: F401
+from .base import ModifierBase  # noqa: F401, isort: skip
+from .atmosphere import CO2Corrector  # noqa: F401
+from .atmosphere import PSPAtmosphericalCorrection  # noqa: F401
+from .atmosphere import PSPRayleighReflectance  # noqa: F401
```
Comment (gerritholl) [inline satpy/modifiers/__init__.py:25]: Same here.  Why make these imports more verbose by repeating the `from .atmosphere import ` part?  Isn't this a violation of DRY?

File: satpy/readers/__init__.py [hunk 6]
```diff
@@ -16,31 +16,33 @@
...
 from functools import total_ordering
-import pickle
 
 import yaml
 
 try:
     from yaml import UnsafeLoader
 except ImportError:
-    from yaml import Loader as UnsafeLoader
+    from yaml import Loader as UnsafeLoader  # type: ignore
```
Comment (gerritholl) [inline satpy/readers/__init__.py:33]: What does the comment `# type: ignore` do?  Is some linter going to complain about this line otherwise?

File: satpy/readers/ahi_hsd.py [hunk 7]
```diff
@@ -33,22 +33,26 @@
...
+from satpy.readers._geos_area import get_area_definition, get_area_extent
+from satpy.readers.file_handlers import BaseFileHandler
+from satpy.readers.utils import (
+    apply_rad_correction,
+    get_earth_radius,
+    get_geostationary_mask,
+    get_user_calibration_factors,
+    np2str,
+    unzip_file,
+)
```
Comment (gerritholl) [inline satpy/readers/ahi_hsd.py:55]: But where it _does_ tolerate having six imports from one module, spread over eight lines.  What's the difference with the cases where it insists on repeating the module the names are imported from?

File: satpy/readers/fci_l1c_nc.py [hunk 8]
```diff
@@ -84,15 +84,15 @@
 .. _test data release: https://www.eumetsat.int/simulated-mtg-fci-l1c-enhanced-non-nominal-datasets
 """
 
-from __future__ import (division, absolute_import, print_function,
-                        unicode_literals)
+from __future__ import absolute_import, division, print_function, unicode_literals
```
Comment (gerritholl) [inline satpy/readers/fci_l1c_nc.py:87]: Not sure why this line was changed, but these future directives are all redundant now.

File: satpy/readers/grib.py [hunk 9]
```diff
@@ -101,9 +102,9 @@ def _create_dataset_ids(self, keys):
             self._msg_datasets[msg_id] = ds_info
 
     @staticmethod
-    def _convert_datetime(msg, date_key, time_key, format="%Y%m%d%H%M"):
+    def _convert_datetime(msg, date_key, time_key, date_format="%Y%m%d%H%M"):
```
Comment (gerritholl) [inline satpy/readers/grib.py:105]: Does a new linter prohibit shadowing builtins?

File: satpy/readers/mviri_l1b_fiduceo_nc.py [hunk 10]
```diff
@@ -153,19 +153,15 @@
...
 import numpy as np
 import xarray as xr
 
 from satpy import CHUNK_SIZE
-from satpy.readers._geos_area import (
-    sampling_to_lfac_cfac,
-    get_area_definition,
-    get_area_extent
-)
+from satpy.readers._geos_area import get_area_definition, get_area_extent, sampling_to_lfac_cfac
```
Comment (gerritholl) [inline satpy/readers/mviri_l1b_fiduceo_nc.py:164]: Why were those changed to be on one line?  Elsewhere the change was the other way around.

File: satpy/tests/reader_tests/test_eps_l1b.py [hunk 11]
```diff
@@ -163,20 +163,16 @@ def mock_getitem(key):
...
-        avhrr_reader.sun_zen = None
-        avhrr_reader.sat_zen = None
         avhrr_reader.scanlines = 7
         avhrr_reader.pixels = 2048
 
         with mock.patch.dict("sys.modules", geotiepoints=geotiemock):
             # Get dask arrays
             sun_azi, sun_zen, sat_azi, sat_zen = avhrr_reader.get_full_angles()
             # Convert to numpy array
-            sun_zen_np1 = np.array(avhrr_reader.sun_zen)
```
Comment (gerritholl) [inline satpy/tests/reader_tests/test_eps_l1b.py:177]: schrödinbug?  How did this ever work in the past?

File: satpy/tests/reader_tests/test_modis_l1b.py [hunk 12]
```diff
@@ -24,13 +24,14 @@
...
-from satpy import available_readers, Scene
+from satpy import Scene, available_readers
+
 from ..utils import CustomScheduler, make_dataid
 from ._modis_fixtures import (
     AVAILABLE_1KM_PRODUCT_NAMES,
     AVAILABLE_HKM_PRODUCT_NAMES,
     AVAILABLE_QKM_PRODUCT_NAMES,
-    _shape_for_resolution
+    _shape_for_resolution,
```
Comment (gerritholl) [inline satpy/tests/reader_tests/test_modis_l1b.py:34]: (When?) do the new linters require a final comma?

File: satpy/tests/reader_tests/test_olci_nc.py [hunk 13]
```diff
@@ -180,8 +183,10 @@ class TestBitFlags(unittest.TestCase):
 
     def test_bitflags(self):
         """Test the BitFlags class."""
-        import numpy as np
         from functools import reduce
+
+        import numpy as np
+
```
Comment (gerritholl) [inline satpy/tests/reader_tests/test_olci_nc.py:189]: Why do we need an empty line after the imports here, but not in line 120/121 or 153/154?

File: satpy/tests/reader_tests/test_slstr_l2.py [hunk 14]
```diff
@@ -19,9 +19,10 @@
 
 import unittest
 from unittest import mock
-from unittest.mock import MagicMock
-from unittest.mock import patch
+from unittest.mock import MagicMock, patch
```
Comment (gerritholl) [inline satpy/tests/reader_tests/test_slstr_l2.py:22]: Here it changed multiple imports to a single one, but elsewhere (see above) it did the exact opposite.  What is the rule?

File: satpy/tests/test_composites.py [hunk 15]
```diff
@@ -90,6 +91,7 @@ def test_mult_ds_diff_area(self):
     def test_mult_ds_diff_dims(self):
         """Test that datasets with different dimensions still pass."""
         from satpy.composites import CompositeBase
+
```
Comment (gerritholl) [inline satpy/tests/test_composites.py:94]: Again; why is there an empty line added in lines 94 and 107, but not between lines 79 and 80 on 179 and 180?

File: satpy/tests/writer_tests/test_mitiff.py [hunk 16]
```diff
@@ -41,9 +41,10 @@ def tearDown(self):
 
     def _get_test_datasets(self):
         """Create a datasets list."""
-        import xarray as xr
-        import dask.array as da
         from datetime import datetime
```
Comment (gerritholl) [inline satpy/tests/writer_tests/test_mitiff.py:44]: I know it's not from this PR, but just for my information — is there a good reason to have a python stdlib import repeated in every test rather than at the module level?

File: satpy/tests/writer_tests/test_ninjotiff.py [hunk 17]
```diff
@@ -101,8 +103,9 @@ def test_convert_units_self(self):
     def test_convert_units_temp(self):
         """Test that temperature unit conversions works as expected."""
         # test converting between °C and K
-        from ..utils import make_fake_scene
         from satpy.writers.ninjotiff import convert_units
+
+        from ..utils import make_fake_scene
```
Comment (gerritholl) [inline satpy/tests/writer_tests/test_ninjotiff.py:108]: So should there be an empty line after imports in a function or not?

File: utils/convert_to_ninjotiff.py [hunk 18]
```diff
@@ -26,14 +26,15 @@
 
 """
 
+import argparse
 import os
 
-from satpy.utils import debug_on
+import yaml
+from mpop.projector import get_area_def
```
Comment (gerritholl) [inline utils/convert_to_ninjotiff.py:33]:
```suggestion
from satpy.pyresample import get_area_def
```

File: .pre-commit-config.yaml [hunk 19]
```diff
@@ -1,8 +1,33 @@
...
+        additional_dependencies:
+          - types-docutils
+          - types-pkg-resources
+          - types-PyYAML
+          - types-requests
+  - repo: https://github.com/pycqa/isort
+    rev: 5.10.1
+    hooks:
+      - id: isort
+        language_version: python3
```
Comment (gerritholl) [inline .pre-commit-config.yaml:29]: What do those commit hooks do exactly?  I suppose it's two more development dependencies.  How do they complement or replace tests currently run in CI?  Do they edit files?  Rewriting commits or adding new commits?  Do they enforce any rules that aren't written either in PEP8 or in Satpy developer guides?

**Review by gerritholl — COMMENTED** [review]
Some questions.  I don't understand entirely the consistency of the changes automatically introduced by (I think) isort, and I don't know how the other newly added hooks work or what they do.

File: satpy/demo/__init__.py (hunk 3 above)
Comment (mraspaud) [reply satpy/demo/__init__.py:54]: might be a line length thing

File: satpy/readers/__init__.py (hunk 6 above)
Comment (mraspaud) [reply satpy/readers/__init__.py:33]: Mypy was complaining. I don't know why though

Comment (djhoese) [reply satpy/readers/__init__.py:33]: Yes, mypy complained and is still complaining in other modules with:

```
Incompatible import of "UnsafeLoader" (imported name has type "Type[Loader]", local name has type "Type[UnsafeLoader]")
```

I'm not sure there is a way around that.

File: satpy/readers/ahi_hsd.py (hunk 7 above)
Comment (djhoese) [reply satpy/readers/ahi_hsd.py:55]: So this may be an inconsistency when isort (the tool reordering the imports) wasn't able to parse the config that @mraspaud had originally made. So in the modifiers file it split them across multiple lines which is against what we configured it for. Or vice versa. It is definitely something we should play with and see what it does. I agree with Martin in the other comment that sometimes I think it is a line length thing.

File: satpy/tests/reader_tests/test_eps_l1b.py (hunk 11 above)
Comment (mraspaud) [reply satpy/tests/reader_tests/test_eps_l1b.py:177]: because the sun_zen attribute in avhrr_reader was filled up by the get_full_angles. Now I'm doing this explicitely

File: satpy/readers/fci_l1c_nc.py (hunk 8 above)
Comment (djhoese) [reply satpy/readers/fci_l1c_nc.py:87]: PR welcome

File: satpy/readers/grib.py (hunk 9 above)
Comment (djhoese) [reply satpy/readers/grib.py:105]: flake8-bugbear I think it was. And yes, it is shadowing builtins.

File: satpy/readers/mviri_l1b_fiduceo_nc.py (hunk 10 above)
Comment (djhoese) [reply satpy/readers/mviri_l1b_fiduceo_nc.py:164]: Maybe they fit within 120 characters?

File: satpy/tests/reader_tests/test_modis_l1b.py (hunk 12 above)
Comment (djhoese) [reply satpy/tests/reader_tests/test_modis_l1b.py:34]: I personally prefer the final comma. FYI isort is currently configured to follow "black" rules when reformatting the import code. This may explain some of the weird behavior of commas versus non-commas and where it breaks newlines.

File: satpy/tests/reader_tests/test_olci_nc.py (hunk 13 above)
Comment (djhoese) [reply satpy/tests/reader_tests/test_olci_nc.py:189]: It is separating the imports into 3 (there are actually 4 or 5) groups: builtins, third party, and first party (satpy).

File: .pre-commit-config.yaml (hunk 19 above)
Comment (djhoese) [reply .pre-commit-config.yaml:29]: Bandit, mypy, and flake8 will only warn about issues. isort is the only one that will rewrite the file. If any of these fail the commit is not done and you have to reattempt the commit. If isort is the only one that failed (meaning it reordered the imports) then in most cases you can just reattempt the commit right away (assuming import reorder doesn't break things).

File: satpy/demo/fci.py (hunk 4 above)
Comment (djhoese) [reply satpy/demo/fci.py:27]: It is not a PEP8 thing and is configured in pyproject.toml for the "isort" tool.
